src/collector.rs

```diff
@@ -325,9 +387,58 @@ impl Collector {
         }
     }
 
+    async fn accept_messages(&mut self, messages: impl Stream<Item = Message>) {
+        futures_util::pin_mut!(messages);
+        while let Some(message) = messages.next().await {
+            let result = async {
+                match message {
+                    Message::Datum(datum) => {
+                        let timestamp = current_timestamp();
+                        let slot = get_timeslot(&mut self.metrics_data, &self.config, timestamp);
+
+                        accept_datum(slot, &mut self.metrics_config, datum);
+
+                        // `current_timestamp` can be pretty expensive when there are a lot of
+                        // metrics so as long as we are immediately able to read more datums we
+                        // assume that we can use the same timestamp for those, thereby amortizing
+                        // the cost.
+                        // 100 is arbitrarily chosen to get good a good amount of reuse without
+                        // risking that this is always ready, thereby never updating the timestamp
+                        for _ in 0..100 {
+                            match future::lazy(|cx| messages.as_mut().poll_next(cx)).await {
+                                Poll::Ready(Some(message)) => match message {
+                                    Message::Datum(datum) => {
+                                        accept_datum(slot, &mut self.metrics_config, datum)
+                                    }
+                                    Message::SendBatch {
+                                        send_all_before,
+                                        emit_sender,
+                                    } => {
+                                        self.accept_send_batch(send_all_before, emit_sender)?;
+                                        break;
+                                    }
+                                },
+                                Poll::Ready(None) => return Ok(()),
+                                Poll::Pending => tokio::task::yield_now().await,
```

Comment:
Doesn't this yield/await mean that we might continue the for loop quite some time later, when we would be well outside of the current time slot?

Reply:
Fixed